uvm/hier/usr/lib/python3/dist-packages/uvm/login_tools.py

```diff
@@ -0,0 +1,239 @@
+import base64
+import sys
+import re
+import pycurl
+import json
+import crypt
+import sys
+import urllib.parse
+from io import StringIO
+#import uvm_login
+
+def get_app_settings_item(a,b):
+    return None
+def get_uvm_settings_item(a,b):
+    return None
+
+try:
+    from uvm.settings_reader import get_app_settings_item
+    from uvm.settings_reader import get_uvm_settings_item
+except ImportError:
+    pass
+
+def valid_login(req, realm, username, password):
+    if realm == 'Administrator':
+        return admin_valid_login(req, realm, username, password)
+    elif realm == 'Reports':
+        if admin_valid_login(req, 'Administrator', username, password, False):
+            return True
+        else:
+            return reports_valid_login(req, realm, username, password)
+    else:
+        return False
+
+def valid_token(req, token):
+    try:
+        uid=None
+        with open('/usr/share/untangle/conf/uid', 'r') as uidfile:
+            uid=uidfile.read().replace('\n', '')
+
+        buffer = StringIO()
+        postdata = json.dumps({ "token": token, "resourceId": uid  })
+
+        curl = pycurl.Curl()
+        curl.setopt( pycurl.POST, 1 )
+        curl.setopt( pycurl.POSTFIELDS, postdata )
+        curl.setopt( pycurl.NOSIGNAL, 1 )
+        curl.setopt( pycurl.CONNECTTIMEOUT, 30 )
+        curl.setopt( pycurl.TIMEOUT, 30 )
+        #curl.setopt( pycurl.URL, "http://54.152.2.165:1337/AuthenticationService/1/CheckTokenAccess")
+        curl.setopt( pycurl.URL, "https://auth.untangle.com/v1/CheckTokenAccess")
+        curl.setopt( pycurl.HTTPHEADER, ["Content-type: application/json", "Accept: application/json", "AuthRequest: 4E6FAB77-B2DF-4DEA-B6BD-2B434A3AE981"])
+        #curl.setopt( pycurl.VERBOSE, True )
+        curl.setopt( pycurl.WRITEDATA, buffer )
+
+        curl.perform()
+
+        body = buffer.getvalue()
+        print(body)
+        return (body == "true")
+    except:
+        return False
+
+def reports_valid_login(req, realm, username, password, log=True):
+    users = get_app_settings_item('reports','reportsUsers')
+    if users == None:
+        return False;
+    if users['list'] == None:
+        return False;
+    for user in users['list']:
+        if user['emailAddress'] != username:
+            continue;
+
+        pw_hash_shadow = user.get('passwordHashShadow')
+        if pw_hash_shadow:
+            if pw_hash_shadow == crypt.crypt(password, pw_hash_shadow):
+                log_login_if_necessary(log, req, username, True, None)
+                return True
+            else:
+                log_login_if_necessary(log, req, username, False, 'P')
+                return False
+        else:
+            pw_hash_base64 = user['passwordHashBase64']
+            pw_hash = base64.b64decode(pw_hash_base64)
+            raw_pw = pw_hash[0:len(pw_hash) - 8]
+            salt = pw_hash[len(pw_hash) - 8:]
+            b = password + salt
+            if raw_pw == hashlib.md5(b.encode('utf-8')).hexdigest():
```

Comment:
Since this doesn't work in admin_valid_login, it probably doesn't work here. This function seems redundant, also.

Reply:
Not sure about `reports_valid_login` at all, but by all means let's simplify the entire thing as much as possible :)